File: Flyweight/java/Flyweight.java
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;

class Shared{
    private String company;
    
    private String position;
    
    public Shared(String company, String position){
        this.company = company;
        this.position = position;
    }
    
    public String getCompany(){
        return company;
    }
    
    public String getPosition(){
        return position;
    }
}

class Unigue{
    private String name;
    
    private String passport;
    
    public Unigue(String name, String passport){
        this.name = name;
        this.passport = passport;
    }
    
    public String getName(){
        return name;
    }
    
    public String getPassport(){
        return passport;
    }
}

class Flyweight{
    private Shared shared;

    public Flyweight(Shared shared){
        this.shared = shared;
    }

    public void process(Unigue unique){
        System.out.println("New data: shared - " + shared.getCompany() + "_" + shared.getPosition()
                + " unique - " + unique.getName() + "_" + unique.getPassport());
    }

    public String getData(){
        return shared.getCompany() + "_" + shared.getPosition();
    }
}

class FlyweightFactory{
    private Hashtable flyweights;

    private String getKey(Shared shared){
        return shared.getCompany() + "_" + shared.getPosition();
    }

    public FlyweightFactory(List<Shared> shareds){
        flyweights = new Hashtable();
        for (Shared shared : shareds)
            flyweights.put(getKey(shared), new Flyweight(shared));
    }

    public Flyweight getFlyweight(Shared shared){
        String key = getKey(shared);
        if (!flyweights.containsKey(key)){
            System.out.println("Flyweight factory: object not found " + key);
            flyweights.put(key, new Flyweight(shared));
        }
        else {
            System.out.println("Flyweight factory: object found " + key);
        }
        return (Flyweight) flyweights.get(key);
    }

    public void listFlyweights(){
        int count = flyweights.size();
        System.out.println("Flyweight factory: count = " + count);
        for (Object pair : flyweights.values()){
            System.out.println(((Flyweight)pair).getData());
        }
    }
}

public class Main {
    static void addSpecialistDatabase(FlyweightFactory flyweightFactory, String company, String position,
                                       String name, String passport){
        System.out.println();
        Flyweight flyweight = flyweightFactory.getFlyweight(new Shared(company, position));
        flyweight.process(new Unigue(name, passport));
    }

    public static void main(String[] args){
        List<Shared> shareds = new ArrayList<>();
        shareds.add(new Shared("Microsoft", "manager"));
        shareds.add(new Shared("Google", "manager"));
        shareds.add(new Shared("Google", "developer"));
        shareds.add(new Shared("Apple", "developer"));

        FlyweightFactory flyweightFactory = new FlyweightFactory(shareds);

        flyweightFactory.listFlyweights();

        addSpecialistDatabase(flyweightFactory,
                "Google",
                "developer",
                "George",
                "12345");

        addSpecialistDatabase(flyweightFactory,
                "Apple",
                "manager",
                "Mark",
                "67890");

        flyweightFactory.listFlyweights();
    }
}

File: Flyweight/c#/Flyweight.cs
using System.Collections;

struct Shared
{
    private string _company;
    
    private string _position;

    public Shared(string company, string position)
    {
        _company = company;
        _position = position;
    }
    
    public string Company { get => _company; }

    public string Position { get => _position; }
}

struct Unique
{
    private string _name;
    
    private string _passport;

    public Unique(string name, string passport)
    {
        _name = name;
        _passport = passport;
    }
    
    public string Name { get => _name; }
    
    public string Passport { get => _passport; }
}

class Flyweight
{
    private Shared _shared;
    
    public Flyweight(Shared shared) => _shared = shared;

    public void Process(Unique unique)
    {
        Console.WriteLine("New data: shared - " + _shared.Company + "_" + _shared.Position + " unique - " + unique.Name + "_" + unique.Passport);
    }

    public string GetData() => _shared.Company + "_" + _shared.Position;
}

class FlyweightFactory
{
    private Hashtable _flyweights;
    
    private string GetKey(Shared shared) => shared.Company + "_" + shared.Position;

    public FlyweightFactory(List<Shared> shareds)
    {
        _flyweights = new Hashtable();
        foreach (Shared shared in shareds)
        {
            _flyweights.Add(GetKey(shared), new Flyweight(shared));
        }
    }

    public Flyweight GetFlyweight(Shared shared)
    {
        string key = GetKey(shared);
        if (!_flyweights.Contains(key))
        {
            Console.WriteLine("Flyweight factory: object not found " + key);
            _flyweights.Add(key, new Flyweight(shared));
        }
        else
        {
            Console.WriteLine("Flyweight factory: object found " + key);
        }

        return (Flyweight) _flyweights[key];
    }

    public void ListFlyweights()
    {
        int count = _flyweights.Count;
        Console.WriteLine("Flyweight factory: count = " + count);
        foreach (Flyweight pair in _flyweights.Values)
        {
            Console.WriteLine(pair.GetData());
        }
    }
}

class Program
{
    static void AddSpecialistDatabase(FlyweightFactory flyweightFactory,
        string company, string position, string name, string passport)
    {
        Console.WriteLine();
        Flyweight flyweight = flyweightFactory.GetFlyweight(new Shared(company, position));
        flyweight.Process(new Unique(name, passport));
    }

    static void Main()
    {
        FlyweightFactory flyweightFactory = new FlyweightFactory(new List<Shared>()
        {
            new Shared("Microsoft", "manager"),
            new Shared("Google", "manager"),
            new Shared("Google", "developer"),
            new Shared("Apple", "developer"),
        });
        
        flyweightFactory.ListFlyweights();
        
        AddSpecialistDatabase(flyweightFactory,
            "Google",
            "developer",
            "George",
            "12345");
        
        AddSpecialistDatabase(flyweightFactory,
            "Apple",
            "manager",
            "Mark",
            "67890");
        
        flyweightFactory.ListFlyweights();
    }
}